File: app/src/main/java/cn/edu/pku/pkurunner/Map/WGS84_GCJ02.java
package cn.edu.pku.pkurunner.Map;

public class WGS84_GCJ02 {
    private static boolean a(double value, double value2) {
        return value < 72.004d || value > 137.8347d || value2 < 0.8293d || value2 > 55.8271d;
    }

    private static double b(double value, double value2) {
        double value3 = value * 2.0d;
        double sqrt = (-100.0d) + value3 + (value2 * 3.0d) + (value2 * 0.2d * value2) + (0.1d * value * value2) + (Math.sqrt(Math.abs(value)) * 0.2d) + ((((Math.sin((6.0d * value) * 3.141592653589793d) * 20.0d) + (Math.sin(value3 * 3.141592653589793d) * 20.0d)) * 2.0d) / 3.0d);
        double d5 = value2 * 3.141592653589793d;
        return sqrt + ((((Math.sin(d5) * 20.0d) + (Math.sin((value2 / 3.0d) * 3.141592653589793d) * 40.0d)) * 2.0d) / 3.0d) + ((((Math.sin((value2 / 12.0d) * 3.141592653589793d) * 160.0d) + (Math.sin(d5 / 30.0d) * 320.0d)) * 2.0d) / 3.0d);
    }

    public static double[] convert(double value, double value2) {
        if (a(value, value2)) {
            return new double[]{value, value2};
        }
        double value3 = value - 105.0d;
        double d5 = value2 - 35.0d;
        double character = c(value3, d5);
        double flag = b(value3, d5);
        double d6 = (value2 / 180.0d) * 3.141592653589793d;
        double sin = Math.sin(d6);
        double d7 = 1.0d - ((0.006693421622965943d * sin) * sin);
        double sqrt = Math.sqrt(d7);
        return new double[]{value + ((character * 180.0d) / (((6378245.0d / sqrt) * Math.cos(d6)) * 3.141592653589793d)), value2 + ((flag * 180.0d) / ((6335552.717000426d / (d7 * sqrt)) * 3.141592653589793d))};
    }

    private static double c(double value, double value2) {
        double value3 = value * 0.1d;
        return value + 300.0d + (value2 * 2.0d) + (value3 * value) + (value3 * value2) + (Math.sqrt(Math.abs(value)) * 0.1d) + ((((Math.sin((6.0d * value) * 3.141592653589793d) * 20.0d) + (Math.sin((value * 2.0d) * 3.141592653589793d) * 20.0d)) * 2.0d) / 3.0d) + ((((Math.sin(value * 3.141592653589793d) * 20.0d) + (Math.sin((value / 3.0d) * 3.141592653589793d) * 40.0d)) * 2.0d) / 3.0d) + ((((Math.sin((value / 12.0d) * 3.141592653589793d) * 150.0d) + (Math.sin((value / 30.0d) * 3.141592653589793d) * 300.0d)) * 2.0d) / 3.0d);
    }
}

File: app/src/main/java/cn/edu/pku/openrunner/feature/run/domain/Wgs84Gcj02.kt
package cn.edu.pku.openrunner.feature.run.domain

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/** Converts GPS WGS-84 coordinates to the GCJ-02 coordinates expected by AMap. */
object Wgs84Gcj02 {
    fun convert(longitude: Double, latitude: Double): Pair<Double, Double> {
        if (outOfChina(longitude, latitude)) return longitude to latitude
        val dLat = transformLatitude(longitude - 105.0, latitude - 35.0)
        val dLon = transformLongitude(longitude - 105.0, latitude - 35.0)
        val radLat = latitude / 180.0 * Math.PI
        val magic = 1 - 0.006693421622965943 * sin(radLat).pow(2)
        val sqrtMagic = sqrt(magic)
        val mgLat = latitude + (dLat * 180.0) /
            ((6335552.717000426 / (magic * sqrtMagic)) * Math.PI)
        val mgLon = longitude + (dLon * 180.0) /
            ((6378245.0 / sqrtMagic * cos(radLat)) * Math.PI)
        return mgLon to mgLat
    }

    private fun outOfChina(longitude: Double, latitude: Double): Boolean =
        longitude < 72.004 || longitude > 137.8347 || latitude < 0.8293 || latitude > 55.8271

    private fun transformLatitude(x: Double, y: Double): Double {
        var ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y +
            0.1 * x * y + 0.2 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * Math.PI) + 20.0 * sin(2.0 * x * Math.PI)) * 2.0 / 3.0
        ret += (20.0 * sin(y * Math.PI) + 40.0 * sin(y / 3.0 * Math.PI)) * 2.0 / 3.0
        ret += (160.0 * sin(y / 12.0 * Math.PI) + 320 * sin(y * Math.PI / 30.0)) * 2.0 / 3.0
        return ret
    }

    private fun transformLongitude(x: Double, y: Double): Double {
        var ret = 300.0 + x + 2.0 * y + 0.1 * x * x +
            0.1 * x * y + 0.1 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * Math.PI) + 20.0 * sin(2.0 * x * Math.PI)) * 2.0 / 3.0
        ret += (20.0 * sin(x * Math.PI) + 40.0 * sin(x / 3.0 * Math.PI)) * 2.0 / 3.0
        ret += (150.0 * sin(x / 12.0 * Math.PI) + 300.0 * sin(x / 30.0 * Math.PI)) * 2.0 / 3.0
        return ret
    }
}
